### PR TITLE
Support "No Load Supply Air Flow Rate Ratio" on UnitarySystemPerformance:Multispeed

### LINKED ISSUE #4747: Support "No Load Supply Air Flow Rate Ratio" on UnitarySystemPerformance:Multispeed
Enhancement Request
--------------

Currently this field neither exists in the OS IDD nor gets forward translated to IDF.

### PR DESCRIPTION
**Pull request by joseph-robertson**
Pull request overview
---------------------

 - Fixes #4747
 - Checklist:
   - [x] Update idd
   - [x] Update model files
   - [x] Update ft file
   - [x] Update model tests
   - [x] Add vt

### Pull Request Author

 - [x] Model API Changes / Additions
 - [ ] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [ ] Model API methods are tested (in `src/model/test`)
 - [ ] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [ ] If a new object or method, added a test in NREL/OpenStudio-resources: Add Link
 - [ ] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [x] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [x] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

File: resources/model/OpenStudio.idd [hunk 1]
```diff
@@ -11361,7 +11361,14 @@ OS:UnitarySystemPerformance:Multispeed,
...
        \note Yes = operate at the highest speed possible without exceeding the current load.
        \note No = allow operation at the average of two adjacent speeds to match the current load.
-  N1 , \field Heating Speed 1 Supply Air Flow Ratio
+  N1 , \field No Load Supply Air Flow Rate Ratio
+       \required-field
+       \type real
+       \minimum 0
+       \maximum 1
+       \note Used to define the no load operating air flow rate when the system fan
+       \note is specified to operate continuously.
```
Comment (jmarrec) [inline resources/model/OpenStudio.idd:11370]: Ok, default of 1 is changed to be a required-field, :+1:

File: src/model/UnitarySystemPerformanceMultispeed.cpp [hunk 2]
```diff
@@ -80,6 +80,17 @@ namespace model {
...
+    double UnitarySystemPerformanceMultispeed_Impl::noLoadSupplyAirflowRateRatio() const {
+      boost::optional<double> value = getDouble(OS_UnitarySystemPerformance_MultispeedFields::NoLoadSupplyAirFlowRateRatio, true);
+      OS_ASSERT(value);
+      return value.get();
+    }
+
+    bool UnitarySystemPerformanceMultispeed_Impl::setNoLoadSupplyAirflowRateRatio(double noLoadSupplyAirflowRateRatio) {
+      bool result = setDouble(OS_UnitarySystemPerformance_MultispeedFields::NoLoadSupplyAirFlowRateRatio, noLoadSupplyAirflowRateRatio);
+      return result;
+    }
```
Comment (jmarrec) [inline src/model/UnitarySystemPerformanceMultispeed.cpp:92]: A new getter and a setter (no reset), all good.

File: src/model/UnitarySystemPerformanceMultispeed.cpp [hunk 3]
```diff
@@ -203,6 +214,9 @@ namespace model {
   UnitarySystemPerformanceMultispeed::UnitarySystemPerformanceMultispeed(const Model& model)
     : ModelObject(UnitarySystemPerformanceMultispeed::iddObjectType(), model) {
     OS_ASSERT(getImpl<detail::UnitarySystemPerformanceMultispeed_Impl>());
+
+    bool ok = setNoLoadSupplyAirflowRateRatio(1.0);
+    OS_ASSERT(ok);
```
Comment (jmarrec) [inline src/model/UnitarySystemPerformanceMultispeed.cpp:219]: Ctor sets the required field, good.

File: src/energyplus/ForwardTranslator/ForwardTranslateUnitarySystemPerformanceMultispeed.cpp [hunk 4]
```diff
@@ -105,6 +105,9 @@ namespace energyplus {
     // Single mode operation
     sysPerf.setString(UnitarySystemPerformance_MultispeedFields::SingleModeOperation, modelObject.singleModeOperation() ? "Yes" : "No");
 
+    // No Load Supply Air Flow Rate Ratio
+    sysPerf.setDouble(UnitarySystemPerformance_MultispeedFields::NoLoadSupplyAirFlowRateRatio, modelObject.noLoadSupplyAirflowRateRatio());
+
```
Comment (jmarrec) [inline src/energyplus/ForwardTranslator/ForwardTranslateUnitarySystemPerformanceMultispeed.cpp:110]: FT sets the field, which is required, so no if optional, :+1:

File: src/model/test/UnitarySystemPerformanceMultispeed_GTest.cpp [hunk 5]
```diff
@@ -66,6 +66,11 @@ TEST_F(ModelFixture, UnitarySystemPerformanceMultispeed_GetterSetters) {
   testObject.resetSingleModeOperation();
   EXPECT_FALSE(testObject.singleModeOperation());
 
+  // No Load
+  EXPECT_EQ(1.0, testObject.noLoadSupplyAirflowRateRatio());
+  EXPECT_TRUE(testObject.setNoLoadSupplyAirflowRateRatio(0.5));
+  EXPECT_EQ(0.5, testObject.noLoadSupplyAirflowRateRatio());
```
Comment (jmarrec) [inline src/model/test/UnitarySystemPerformanceMultispeed_GTest.cpp:72]: No probleme here

File: src/energyplus/ForwardTranslator/ForwardTranslateUnitarySystemPerformanceMultispeed.cpp (hunk 4 above)
Comment (jmarrec) [reply src/energyplus/ForwardTranslator/ForwardTranslateUnitarySystemPerformanceMultispeed.cpp:110]: ideally a FT test would ensure this works appropriately.

File: src/osversion/VersionTranslator.cpp [hunk 6]
```diff
@@ -7650,9 +7651,50 @@ namespace osversion {
...
+
+        // 1 Field has been added from 3.5.0 to 3.5.1:
+        // ----------------------------------------------
+        // * No Load Supply Air Flow Rate Ratio * 3
+        auto iddObject = idd_3_5_1.getObject(iddname);
+        IdfObject newObject(iddObject.get());
+
+        for (size_t i = 0; i < object.numFields(); ++i) {
+          auto value = object.getString(i);
+          if (value) {
```
Comment (jmarrec) [inline src/osversion/VersionTranslator.cpp]:
```suggestion
          if ((value = object.getString(i))) {
```

File: src/osversion/VersionTranslator.cpp [hunk 7]
```diff
@@ -7650,9 +7651,50 @@ namespace osversion {
...
+            } else {
+              newObject.setString(i + 1, value.get());
+            }
+          }
+        }
+
+        newObject.setDouble(3, 1.0);
+
+        m_refactored.push_back(RefactoredObjectData(object, newObject));
+        ss << newObject;
```
Comment (jmarrec) [inline src/osversion/VersionTranslator.cpp:7686]: No problem here, all good.

File: src/osversion/test/VersionTranslator_GTest.cpp [hunk 8]
```diff
@@ -2392,3 +2392,28 @@ TEST_F(OSVersionFixture, update_3_4_0_to_3_5_0_TableMultiVariableLookup_oneDim_o
...
+  openstudio::path outPath = resourcesPath() / toPath("osversion/3_5_1/test_vt_UnitarySystemPerformanceMultispeed_updated.osm");
+  model->save(outPath, true);
+
+  std::vector<WorkspaceObject> perfs = model->getObjectsByType("OS:UnitarySystemPerformance:Multispeed");
+  ASSERT_EQ(1u, perfs.size());
+  WorkspaceObject perf = perfs[0];
+
+  EXPECT_EQ("Unitary System Performance Multispeed 1", perf.getString(OS_UnitarySystemPerformance_MultispeedFields::Name).get());
+  EXPECT_EQ("No", perf.getString(OS_UnitarySystemPerformance_MultispeedFields::SingleModeOperation).get());
+  EXPECT_EQ(1.0, perf.getDouble(OS_UnitarySystemPerformance_MultispeedFields::NoLoadSupplyAirFlowRateRatio).get());
```
Comment (jmarrec) [inline src/osversion/test/VersionTranslator_GTest.cpp]: Undefined OS_UnitarySystemPerformance_MultispeedFields

**Review by jmarrec — APPROVED** [review]
Made a couple of very minor changes. All good. Once CI reports success, we can merge. Thanks @joseph-robertson

**Comment by jmarrec** [comment]
@joseph-robertson I don't have jenkins access here, could you check why the Windows runner is failing please?

File: src/osversion/test/VersionTranslator_GTest.cpp (hunk 8 above)
Comment (joseph-robertson) [reply src/osversion/test/VersionTranslator_GTest.cpp]: Hmm, these weren't working?

Comment (jmarrec) [reply src/osversion/test/VersionTranslator_GTest.cpp]: unless you #include the specific Object_FieldEnums.hxx it doesn't no. I'm torn about whether we want to start using the FieldEnums or not... it's going to create problems where we need to update them even if it's just the casing that changes, so I chose not to

**Pull request merged**